### PR TITLE
[MLIR][LLVM] Enable export of DISubprograms on function declarations

### PR DESCRIPTION
**Pull request by Dinistro**
This commit changes the MLIR to LLVMIR export to also attach subprogram debug attachements to function declarations.
This commit additonally fixes the two passes that produce subprograms to not attach the "Definition" flag to function declarations. This otherwise results in invalid LLVM IR.

### CODE REVIEW

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir [hunk 1]
```diff
@@ -169,6 +169,20 @@ llvm.func @empty_types() {
...
+#di_subprogram = #llvm.di_subprogram<
+  scope = #di_file, name = "func_decl_with_subprogram", file = #di_file, subprogramFlags = "Optimized"
+>
+
+// CHECK-LABEL: declare !dbg
+// CHECK-SAME: ![[SUBPROGRAM:.*]] i32 @func_decl_with_subprogram(
+llvm.func @func_decl_with_subprogram() -> (i32) loc(fused<#di_subprogram>["foo.mlir":2:1])
+
+// CHECK: ![[SUBPROGRAM]] = !DISubprogram(name: "func_decl_with_subprogram", scope: ![[FILE:.*]], file: ![[FILE]], spFlags: DISPFlagOptimized)
+// CHECK: ![[FILE:.*]] = !DIFile(filename: "foo.mlir", directory: "/test/")
```
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:182]:
```suggestion
// CHECK: ![[FILE]] = !DIFile(filename: "foo.mlir", directory: "/test/")
```
Shouldn't this match the FILE variable matched above?

**Review by gysit — COMMENTED** [review]
Nice! 

It seems like there is a flang test that needs adaption (c.f. CI).

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir (hunk 1 above)
Comment (Dinistro) [reply mlir/test/Target/LLVMIR/llvmir-debug.mlir:182]: Moved it around and didn't update it properly, good catch.

**Review by gysit — APPROVED** [review]
Thanks LGTM!

**Pull request merged**